wallet-next/src/worker.rs

```diff
@@ -1,24 +1,71 @@
-use crate::Storage;
-use penumbra_proto::client::oblivious::oblivious_query_client::ObliviousQueryClient;
+use crate::{sync::scan_block, Storage};
+use penumbra_crypto::{merkle::NoteCommitmentTree, FullViewingKey};
+use penumbra_proto::client::oblivious::{
+    oblivious_query_client::ObliviousQueryClient, CompactBlockRangeRequest,
+};
 use tonic::transport::Channel;
 
 pub struct Worker {
     storage: Storage,
     client: ObliviousQueryClient<Channel>,
-    // TODO: notifications (see TODOs on WalletService)
+    nct: NoteCommitmentTree,
+    fvk: FullViewingKey, // TODO: notifications (see TODOs on WalletService)
 }
 
 impl Worker {
-    pub fn new(storage: Storage, client: ObliviousQueryClient<Channel>) -> Self {
-        Self { storage, client }
+    pub async fn new(
+        storage: Storage,
+        client: ObliviousQueryClient<Channel>,
+    ) -> Result<Self, anyhow::Error> {
+        let nct = storage.note_commitment_tree().await?;
+        let fvk = storage.full_viewing_key().await?;
+        Ok(Self {
+            storage,
+            client,
+            nct,
+            fvk,
+        })
     }
 
-    pub async fn sync_to_latest(&self) -> Result<u64, anyhow::Error> {
+    pub async fn sync_to_latest(&mut self) -> Result<u64, anyhow::Error> {
         // Do a single sync run, up to whatever the latest block height is
-        todo!()
+        tracing::info!("starting client sync");
+
+        let start_height = self
+            .storage
+            .last_sync_height()
+            .await?
+            .map(|h| h + 1)
+            .unwrap_or(0);
+
+        let mut end_height = start_height;
+
+        let mut stream = self
+            .client
+            .compact_block_range(tonic::Request::new(CompactBlockRangeRequest {
+                start_height,
+                end_height: 0,
+                chain_id: self.storage.chain_params().await?.chain_id,
+            }))
+            .await?
+            .into_inner();
+
+        while let Some(block) = stream.message().await? {
+            let scan_result = scan_block(&self.fvk, &mut self.nct, block.try_into()?);
+
+            end_height = std::cmp::max(end_height, scan_result.height);
```

Comment:
The intent here is to ratchet the local `end_height` up to the highest witnessed block height, but as a sanity check, is there any possibility that this could result in missed blocks of a lower height which would only come in later?

Reply:
Yeah, but this problem is much worse -- if we miss blocks, we'll get a corrupted state, because we could miss notes that we should have seen, or else miss spends (e.g., if we would have got a note in block H, and we miss it, and it's spent in block H+K, then when we scan block H+K we won't see that it's something whose spend we should record, even if we scan block H after the fact).

There's no way around the fact that we need to apply each delta update (block) in sequence.

As a sanity check, I think it would make sense to have `record_block` check against `sync_height: Option<u64>` like so:
```rust
let correct_height = match sync_height {
    // Require that the new block follows the last one we scanned.
    Some(cur_height) => scan_result.height == cur_height + 1,
    // Require that the new block represents the initial chain state.
    None => scan_result.height == 0,
};

if !correct_height {
    return Err(anyhow::anyhow!("Wrong block height {:?} for sync height {:?}", scan_result.height, sync_height));
}
```

Reply:
Cool, this makes sense - then presumably just increment the latest height stored in sqlite by 1?

I'm mostly concerned about the role the sqlite `sync_height` table is playing here and want to take extra care to ensure it isn't incremented in such a way that we could end up with corrupted state.

Are there additional checks or mitigations we could be doing, or additional means of verifying we are storing the correct latest height in sqlite?